File: png.py

import zlib

def paeth_predictor(a, b, c):
    pa = abs(b - c)
    pb = abs(a - c)
    pc = abs(a + b - 2*c)
    if pa <= pb and pa <= pc:
        return a
    elif pb <= pc:
        return b
    else:
        return c

class PNG:

    def __init__(self):
        self.image_data = bytearray()
        self.reconstructed = bytearray()
        self.chunkmap = {b'IHDR': self.ihdr_reader,
                         b'sRGB': self.srgb_reader,
                         b'sBIT': self.sbit_reader,
                         b'IDAT': self.idat_reader}
        self.filter_types = []

    def process(self, file):
        signature = file.read(8);

        while True:
            length = int.from_bytes(file.read(4), byteorder='big')
            chunk_type = file.read(4)
            data = file.read(length)
            crc = int.from_bytes(file.read(4), byteorder='big')

            if chunk_type == b'IEND':
                break

            if crc != zlib.crc32(chunk_type + data):
                print(chunk_type)
                print("CRC failed", crc, crc32(data, crc32(data)))
                return False

            if not chunk_type in self.chunkmap:
                print(chunk_type)
                return False

            if not self.chunkmap[chunk_type](length, data):
                print(chunk_type)
                print("error reading chunk")

        self.image_data = zlib.decompress(self.image_data)
        self.reconstruct_png(self.image_data)

        return True

    def reconstruct_png(self, original):
        scanlines_size = self.width*4
        scanlines_and_filter_size = scanlines_size + 1

        for i in range(self.height):
            filter_type = self.image_data[scanlines_and_filter_size * i]
            for j in range(scanlines_size):
                x = self.image_data[scanlines_and_filter_size * i + j + 1]

                a = 0
                b = 0
                c = 0

                if j >= 4:
                    a = self.reconstructed[i*scanlines_size + j - 4]
                if i > 0:
                    b = self.reconstructed[(i-1)*scanlines_size + j]
                if i > 0 and j >= 4:
                    c = self.reconstructed[(i-1)*scanlines_size + j-4]

                if filter_type == 1:
                    x += a
                elif filter_type == 2:
                    x += b
                elif filter_type == 3:
                    x += (a + b)//2
                elif filter_type == 4:
                    x += paeth_predictor(a, b, c)

                self.reconstructed.append(x % 256)

    def difference(self, other):
        differences = []
        scanlines_size = self.width*4
        for i, (b1, b2) in enumerate(zip(self.reconstructed, other.reconstructed)):
            if i % scanlines_size == 0:
                scanline = i//scanlines_size
                differences.append([])
            if b1 != b2:
                pos = (i-scanline*scanlines_size).to_bytes(2, byteorder='big')
                b1 = b1.to_bytes(1, byteorder='big')
                differences[scanline].append((pos, b1))

        return differences

    def ihdr_reader(self, length, data):
        self.width = int.from_bytes(data[0:4], byteorder='big')
        self.height = int.from_bytes(data[4:8], byteorder='big')
        self.bit_depth = data[8]
        self.colour_type = data[9]
        self.compression_method = data[10]
        self.filter_method = data[11]
        self.interlace_method = data[12]

        if self.bit_depth != 8:
            return False

        if self.colour_type != 6:
            return False

        if self.compression_method != 0:
            return False

        if self.interlace_method != 0:
            return False

        return True

    def srgb_reader(self, length, data):
        return True

    def sbit_reader(self, length, data):
        return True

    def idat_reader(self, length, data):
        self.image_data += data

        return True

File: png.cpp

#include "png.h"
#include "crc_utils.h"

#include <QtEndian>
#include <QDebug>

uint8_t paeth_predictor(uint8_t a, uint8_t b, uint8_t c)
{
    int pa = abs(b - c);
    int pb = abs(a - c);
    int pc = abs(a + b - c - c);
    int Pr;
    if ((pa <= pb) && (pa <= pc)) {
        Pr = a;
    } else if (pb <= pc) {
        Pr = b;
    } else {
        Pr = c;
    }
    return (uint8_t) Pr;
}

PNG::PNG() {
    chunkmap = initChunkMap();
}

QMap<QByteArray, PNG::chunk_reader>
PNG::initChunkMap() {
    QMap<QByteArray, PNG::chunk_reader> map;
    PNG_INSERT_LABEL(IHDR);
    PNG_INSERT_LABEL(sRGB);
    PNG_INSERT_LABEL(sBIT);
    PNG_INSERT_LABEL(IDAT);
    return map;
}

bool
PNG::process(QByteArray file_data) {
    uint index = 8;
    QByteArray signature = file_data.left(8);

    while (true) {
        uint32_t length = qFromBigEndian<quint32>(file_data.mid(index, 4).data());
        index += 4;
        QByteArray chunk_type = file_data.mid(index, 4);
        index += 4;
        QByteArray data = file_data.mid(index, length);
        index += length;
        uint32_t crc = qFromBigEndian<quint32>(file_data.mid(index, 4).data());
        index += 4;

        if (chunk_type == IEND)
            break;

        if (crc != crc32(chunk_type + data)) {
            qDebug() << chunk_type;
            qDebug() << "CRC failed" << crc << crc32(data, crc32(data));
            return false;
        }

        if (!chunkmap.contains(chunk_type)) {
            qDebug() << chunk_type;
            return false;
        }

        PNG::chunk_reader reader = chunkmap[chunk_type];
        if(!(this->*reader)(length, data)){
            qDebug() << chunk_type;
            qDebug() << "error reading chunk";
        }
    }

    uint32_t length;
    qToBigEndian<quint32>(image_data.size(), &length);

    qDebug() << "Finished reading";

    image_data = QByteArray((const char *) &length, 4) + image_data;
    image_data = qUncompress(image_data);

    qDebug() << "Finished uncompressing";

    image_data = reconstruct_png(image_data);

    qDebug() << "Finished reconstructing";

    return true;
}

QByteArray
PNG::reconstruct_png(QByteArray &original) {
    uint scanlines_size = width*4;
    uint scanlines_and_filter_size = scanlines_size + 1;
    QByteArray reconstructed;

    for (uint i = 0; i < height; i++) {
        uint row_original = scanlines_and_filter_size * i;
        uint row_reconstructed = scanlines_size * i;
        uint row_up = scanlines_size * (i-1);
        uint8_t filter_type = original[row_original];
        for (uint j = 0; j < scanlines_size; j++) {
            uint8_t x = original[row_original + j + 1];

            uint8_t a = 0;
            uint8_t b = 0;
            uint8_t c = 0;

            if (j >= 4)
                a = (uint8_t) reconstructed[row_reconstructed + j - 4];
            if ((i > 0) && (filter_type > 1))
                b = (uint8_t) reconstructed[row_up + j];
            if ((filter_type == 4) && (i > 0) && (j >= 4))
                c = (uint8_t) reconstructed[row_up + j - 4];

            switch(filter_type) {
                case 1:
                    x += a;
                    break;
                case 2:
                    x += b;
                    break;
                case 3:
                    x += ((a + b)/2);
                    break;
                 case 4:
                    x += paeth_predictor(a, b, c);
                    break;
            }

            reconstructed.append(x);
        }
    }
    return reconstructed;
}


bool
PNG::change_bytes(uint16_t scanline, QByteArray differences) {
    int scanline_offset = width*4*scanline;
    for (int j = 0; j < differences.size(); j += 3) {
        uint32_t pos = qFromBigEndian<quint16>(differences.mid(j, 2).data());
        image_data[scanline_offset + pos] = differences[j+2];
    }

    return true;
}

bool
PNG::change_byte(uint16_t scanline, uint32_t pos, uint8_t byte) {
    int scanlines_size = width*4;
    image_data[scanline*scanlines_size + pos] = byte;

    return true;
}

QColor
PNG::getPixel(int x, int y) {
    int index = width*4*y + x*4;
    uint8_t r = image_data[index    ];
    uint8_t g = image_data[index + 1];
    uint8_t b = image_data[index + 2];
    uint8_t a = image_data[index + 3];

    return QColor(r, g, b, a);
}

bool
PNG::IHDRReader(uint32_t length, QByteArray &data) {
    Q_UNUSED(length);

    width = qFromBigEndian<quint32>(data.mid(0, 4).data());
    height = qFromBigEndian<quint32>(data.mid(4, 4).data());
    bit_depth = *data.mid(8, 1).data();
    colour_type = *data.mid(9, 1).data();
    compression_method = *data.mid(10, 1).data();
    filter_method = *data.mid(11, 1).data();
    interlace_method = *data.mid(12, 1).data();

    if(bit_depth != 8)
        return false;

    if(colour_type != 6)
        return false;

    if(compression_method != 0)
        return false;

    if(interlace_method != 0)
        return false;

    return true;
}

bool
PNG::sRGBReader(uint32_t length, QByteArray &data) {
    Q_UNUSED(length);
    Q_UNUSED(data);

    return true;
}

bool
PNG::sBITReader(uint32_t length, QByteArray &data) {
    Q_UNUSED(length);
    Q_UNUSED(data);

    return true;
}

bool
PNG::IDATReader(uint32_t length, QByteArray &data) {
    Q_UNUSED(length);

    image_data += data;

    return true;
}
